lib/projfs.c

```diff
@@ -1156,6 +1157,126 @@ static int check_dir_empty(const char *path)
 	}
 }
 
+/**
+ * Uses the f*xattr family of functions to check that xattrs are supported on
+ * the supplied open file descriptor.
+ */
+static int test_xattr_support(int fd)
+{
+	int res;
+	char v = 1;
+
+	res = fsetxattr(fd, USER_PROJECTION_TEST, &v, 1, 0);
+	if (res == -1)
+		return -1;
+
+	res = fgetxattr(fd, USER_PROJECTION_TEST, &v, 1);
+	if (res != 1 || v != 1)
+		return -1;
+
+	res = fremovexattr(fd, USER_PROJECTION_TEST);
+	if (res != 0)
+		return -1;
+
+	return 0;
+}
+
+/**
+ * Sets the empty xattr on the given fd, and checks that it was set
+ * successfully. */
+static int set_and_check_empty_xattr(int fd)
+{
+	char v = 1;
+	int res;
+
+	res = fsetxattr(fd, USER_PROJECTION_EMPTY, &v, 1, 0);
+	if (res == -1)
+		return -1;
+
+	res = fgetxattr(fd, USER_PROJECTION_EMPTY, &v, 1);
+	if (res != 1 || v != 1)
+		return -1;
+
+	return 0;
+}
+
+#define SPARSE_TEST_FILENAME ".libprojfs-sparse-test"
+#define SPARSE_TEST_SIZE_BYTES 1048576
+
+/**
+ * Test that using ftruncate to create a sparse file works.
+ */
+static int test_sparse_support(struct projfs *fs)
+{
+	struct stat attrs;
+	int fd, res;
+
+	fd = openat(
+		fs->lowerdir_fd,
+		SPARSE_TEST_FILENAME,
+		O_CREAT | O_TRUNC | O_RDWR,
+		0600);
```

Comment:
This seems to harm rather than aid readability. I feel like they should go one one line all together (optimise for compactness at the expense of readability), or separate lines (such that line X of the parameter list corresponds to parameter X, optimising for readability at the expense of compactness). Mixing the styles appears to yield the worst of both worlds (both difficult to read but not particularly compact).

Reply:
For consistency's sake, I'll do it anyway!